### PR TITLE
group and simplify iptables INPUT rules

### LINKED ISSUE #1236: Redundant iptables rules?
Looking over the iptables ruleset it seems that [this rule](https://github.com/freedomofpress/securedrop/blob/develop/install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4#L24) on line 24:
`-A INPUT -p tcp -m state --state ESTABLISHED,RELATED -j ACCEPT -m comment --comment "Allow traffic back for tor"`
shadows the later rules [on line 66](https://github.com/freedomofpress/securedrop/blob/develop/install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4#L66) and [line 89](https://github.com/freedomofpress/securedrop/blob/develop/install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4#L89):
```
-A INPUT -p tcp --match multiport --sports 80,8080,443 -m state --state ESTABLISHED,RELATED -j ACCEPT -m comment --comment "apt updates"
-A INPUT -p tcp --sport {{ smtp_relay_port }} -m state --state ESTABLISHED,RELATED -j ACCEPT -m comment --comment "Allow ossec email alerts out"
```
I haven't configured a local install for testing, but I would guess that the latter rules are never hit?

### PR DESCRIPTION
**Pull request by singuliere**
## Status

Ready for review

## Description of Changes

Fixes #1236

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

### If you made changes to the app code:

- [ ] Unit and functional tests pass on the development VM

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made changes to documentation:

- [ ] Doc linting passed locally

### CODE REVIEW

File: install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4 [hunk 1]
```diff
@@ -4,9 +4,12 @@
 :OUTPUT DROP [0:0]
 :LOGNDROP - [0:0]
 
+-A INPUT         -m state --state ESTABLISHED -j ACCEPT   -m comment --comment "Allow traffic back"
```
[the remark this answers is not in the document]
Comment (emkll) [reply install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: I see this rule omits `-p tcp`, thus allowing all established UDP traffic inbound. While is it less restrictive than the previous iteration, I don't immediately see a problem with this.

File: install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4 [hunk 2]
```diff
@@ -31,62 +33,50 @@
...
 
 # Block all other traffic by application users
 # Load before generic loopback rules
 -A OUTPUT -m owner --uid-owner www-data -j LOGNDROP -m comment --comment "Drop all other traffic by the securedrop user"
 
 {% endif %}
 
-# Block all other outbound access for users in the ssh group
-# Load before generic loopback rules
--A OUTPUT -m owner --gid-owner ssh -j LOGNDROP -m comment --comment "Drop all other outbound traffic for ssh user"
```
Comment (emkll) [inline install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: Why was this line removed?

**Comment by ageis** [comment]
Thanks @singuliere.

Entirely tangential comments that probably deserve their own issue... but about the way SecureDrop manages iptables rules in general. It's not ideal... James Dolan (RIP) left a comment in his commit that the whole process should be redesigned (see head of `install_files/ansible-base/roles/restrict-direct-access/tasks'iptables.yml`)... To begin with, we apply the rules with an ifup-d wrapper, e.g. when a network link becomes active. I can think of ways this could be bypassed. A slight improvement (to the way we apply them) would be to manage them with a systemd service instead. You could accomplish it all in one unit:
```
[Unit]
Description=Manage iptables rules
Before=network.target
Before=shutdown.target
DefaultDependencies=yes

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStartPre=/bin/mkdir -p /etc/network/iptables/
ExecStart=/sbin/iptables-restore /etc/network/iptables/rules_v4
ExecStop=/sbin/iptables-save /etc/network/iptables/rules_v4
ExecReload=/sbin/iptables-save /etc/network/iptables/rules_v4

[Install]
WantedBy=multi-user.target
```

Rules are restored on boot, before there's even networking, and saved every time the host gets shut down. More secure, but also stateful — remembers alterations. This doesn't address the needed bypass in staging/during installation though. We also have issue #845 left open. It would serve well to audit and test the rules to ensure they do what they're supposed to do, or perhaps consider other approaches, such as a firewall management daemon like ufw or nufw/UFWI or strongwall? Just wanted to attempt to kickstart a discussion somewhere, since this piece of the project hasn't seen much love.

**Comment by dachary** [comment]
@singuliere gentle ping?

File: install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4 (hunk 1 above)
[the remark this answers is not in the document]
Comment (singuliere) [reply install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: > why this rule was moved to the top

Because it is generic. It is not limited to tor traffic, contrary to what the comment claims.

File: install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4 [hunk 3]
```diff
@@ -4,9 +4,12 @@
 :OUTPUT DROP [0:0]
 :LOGNDROP - [0:0]
 
+-A INPUT         -m state --state ESTABLISHED -j ACCEPT   -m comment --comment "Allow traffic back"
+
 # Prod ssh connections happen through an authenticated tor hidden service
 # The ssh connection is proxied on the server by the tor client to
 # the ssh dameon listening on the local loopback.
+-A INPUT -i lo -p tcp --dport 22 -m state --state NEW -j ACCEPT -m comment --comment "Allow tor connection from local loopback to connect to ssh"
```
[the remark this answers is not in the document]
Comment (singuliere) [reply install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: this rule is for the INPUT chain

File: install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4 (hunk 2 above)
Comment (singuliere) [reply install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: there are no users in the ssh group

Comment (emkll) [reply install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: Ah, yes!

File: install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4 (hunk 3 above)
[the remark this answers is not in the document]
Comment (emkll) [reply install_files/ansible-base/roles/restrict-direct-access/templates/rules_v4]: Makes sense as the rule on L115 was removed :)

**Comment by dachary** [comment]
@emkll @singuliere IMHO it would make more sense to verify the intended side effect of the iptables rules (e.g. a packet sent to 8080 from the wrong user or interface is actually dropped & logged). There is very little value (if any) in dumping the rules to verify they are identical to the one that were just loaded. At best it verifies iptables-save/restore works as intended. My 2cts ;-)

**Comment by conorsch** [comment]
> it would make more sense to verify the intended side effect of the iptables rules (e.g. a packet sent to 8080 from the wrong user or interface is actually dropped & logged)

:+1: Could not agree more, @dachary. The current config tests, particularly on the host-level firewall side of things, merely vet the intended implementation, rather than the desired system end state. Once we move to externalizing the tests, e.g. using client outside the app server as described in #2875, we should also revisit the config tests and make sure that we're validating our hardening by banging on the server from the outside.

**Comment by singuliere** [comment]
I will write tests

**Comment by dachary** [comment]
@singuliere ping?

**Comment by singuliere** [comment]
I will rebase

**Comment by dachary** [comment]
This needs rebasing.

**Comment by eloquence** [comment]
Hey @singuliere , do you still have time to work on this? If not, we can rebase/re-do the change. Thanks for your initial stab at it!

**Comment by singuliere** [comment]
Closing because I do not have time to work on this.
